File: forms/obstetrics-assessment/back-end-with-loco/src/engine/antenatal_grader.rs
//! Antenatal grader module.

use super::flagged_issues::detect_additional_flags;
use super::ng201_rules::all_rules;
use super::types::{AssessmentData, FiredRule, GradingResult, RiskLevel};

/// Pure function: grades an obstetrics assessment per NICE NG201.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let fired_rules = run_rules(data);
    let additional_flags = detect_additional_flags(data);

    let risk_level: RiskLevel = if fired_rules.iter().any(|r| r.risk == "high") {
        "high".to_string()
    } else if fired_rules.iter().any(|r| r.risk == "moderate") {
        "moderate".to_string()
    } else {
        "low".to_string()
    };

    let answered_count = count_answered_inputs(data);

    GradingResult {
        risk_level,
        answered_count,
        fired_rules,
        additional_flags,
        timestamp,
    }
}

/// Evaluate every NG201 rule against the data, returning the fired rules.
pub fn run_rules(data: &AssessmentData) -> Vec<FiredRule> {
    let mut out: Vec<FiredRule> = Vec::new();
    for rule in all_rules() {
        if let Some(risk) = (rule.evaluate)(data) {
            out.push(FiredRule {
                id: rule.id.to_string(),
                category: rule.category.to_string(),
                description: rule.description.to_string(),
                risk: risk.to_string(),
            });
        }
    }
    // Sort: high first, then moderate, then low; tie-break by ID.
    out.sort_by(|a, b| {
        let order = |r: &str| match r {
            "high" => 0,
            "moderate" => 1,
            "low" => 2,
            _ => 3,
        };
        order(a.risk.as_str())
            .cmp(&order(b.risk.as_str()))
            .then_with(|| a.id.cmp(&b.id))
    });
    out
}

/// Count answered top-level questionnaire inputs across every section as a
/// proxy for completeness. JSON-equivalent: any value that is neither null
/// nor the empty string.
fn count_answered_inputs(data: &AssessmentData) -> u32 {
    let v = match serde_json::to_value(data) {
        Ok(v) => v,
        Err(_) => return 0,
    };
    let mut count = 0u32;
    if let serde_json::Value::Object(sections) = v {
        for (_section_key, section_value) in sections.iter() {
            if let serde_json::Value::Object(fields) = section_value {
                for (_field_key, field_value) in fields.iter() {
                    if is_answered(field_value) {
                        count += 1;
                    }
                }
            }
        }
    }
    count
}

fn is_answered(v: &serde_json::Value) -> bool {
    match v {
        serde_json::Value::Null => false,
        serde_json::Value::String(s) => !s.is_empty(),
        _ => true,
    }
}

/// Friendly label for a risk level.
pub fn risk_level_label(level: &str) -> String {
    match level {
        "high" => "High Risk — Consultant-led / Multidisciplinary Care".to_string(),
        "moderate" => "Moderate Risk — Obstetric Input at Milestones".to_string(),
        "low" => "Low Risk — Midwifery-led Care".to_string(),
        _ => String::new(),
    }
}

/// CSS class hint for a risk badge.
pub fn risk_level_class(level: &str) -> String {
    match level {
        "high" => "risk-high".to_string(),
        "moderate" => "risk-moderate".to_string(),
        "low" => "risk-low".to_string(),
        _ => String::new(),
    }
}

File: forms/obstetrics-assessment/front-end-with-html/js/antenatal-grader.js
import { ng201Rules } from './rules.js';

// NICE NG201 Antenatal Risk grader. Pure functions: take an
// `AssessmentData` object, evaluate every NG201 rule, and return the
// overall risk level plus the audit trail of fired rules.
//
// Risk stratification:
//   any rule fires 'high'      -> overall 'high'      (consultant-led care)
//   else any 'moderate' fires  -> overall 'moderate'  (obstetric input at milestones)
//   otherwise                  -> overall 'low'       (midwifery-led care)

/**
 * @typedef {import('./types.js').AssessmentData} AssessmentData
 * @typedef {import('./types.js').RiskLevel} RiskLevel
 * @typedef {import('./types.js').FiredRule} FiredRule
 */

// Wrapped in an IIFE; published via window.ObstetricsAssessment.

/**
 * Friendly label for a RiskLevel.
 * @param {RiskLevel} level
 */
function riskLevelLabel(level) {
  switch (level) {
    case 'high': return 'High Risk — Consultant-led / Multidisciplinary Care';
    case 'moderate': return 'Moderate Risk — Obstetric Input at Milestones';
    case 'low': return 'Low Risk — Midwifery-led Care';
    default: return '';
  }
}

/**
 * CSS class hint for the risk badge.
 * @param {RiskLevel} level
 */
function riskLevelClass(level) {
  switch (level) {
    case 'high': return 'risk-high';
    case 'moderate': return 'risk-moderate';
    case 'low': return 'risk-low';
    default: return '';
  }
}

/**
 * Evaluate every NG201 rule against the supplied assessment data and
 * combine the per-rule risk levels into an overall risk stratification.
 *
 * @param {AssessmentData} data
 * @returns {{ riskLevel: RiskLevel, answeredCount: number, firedRules: FiredRule[] }}
 */
function calculateAntenatalRisk(data) {
  /** @type {FiredRule[]} */
  const firedRules = [];

  for (const rule of ng201Rules) {
    try {
      const risk = rule.evaluate(data);
      if (risk) {
        firedRules.push({
          id: rule.id,
          category: rule.category,
          description: rule.description,
          risk
        });
      }
    } catch (e) {
      console.warn(`NG201 rule ${rule.id} evaluation failed:`, e);
    }
  }

  /** @type {RiskLevel} */
  let riskLevel = 'low';
  if (firedRules.some((r) => r.risk === 'high')) {
    riskLevel = 'high';
  } else if (firedRules.some((r) => r.risk === 'moderate')) {
    riskLevel = 'moderate';
  }

  // Sort fired rules: high first, then moderate, then low (by id within bucket).
  const order = { high: 0, moderate: 1, low: 2 };
  firedRules.sort((a, b) => order[a.risk] - order[b.risk] || a.id.localeCompare(b.id));

  // Count answered fields used as primary inputs (a proxy for completeness).
  const answeredCount = countAnsweredInputs(data);

  return { riskLevel, answeredCount, firedRules };
}

/**
 * Count answered top-level questionnaire inputs to provide a denominator
 * the report can show ("X relevant fields answered").
 * @param {AssessmentData} data
 */
function countAnsweredInputs(data) {
  let count = 0;
  const isAnswered = (v) => v !== null && v !== undefined && v !== '';
  for (const section of Object.keys(data)) {
    for (const key of Object.keys(data[section])) {
      if (isAnswered(data[section][key])) count++;
    }
  }
  return count;
}

export { riskLevelLabel, riskLevelClass, calculateAntenatalRisk };
